File: python/ids.py
from __future__ import annotations

import threading
import uuid
from typing import Protocol

class IdSource(Protocol):
    def session(self) -> str: ...
    def item(self) -> str: ...
    def response(self) -> str: ...
    def event(self) -> str: ...
    def turn(self) -> str: ...
    def phrase(self) -> str: ...

class RandomIdSource:
    def session(self) -> str:
        return f"sess_{uuid.uuid4().hex}"

    def item(self) -> str:
        return f"item_{uuid.uuid4().hex}"

    def response(self) -> str:
        return f"resp_{uuid.uuid4().hex}"

    def event(self) -> str:
        return f"evt_{uuid.uuid4().hex}"

    def turn(self) -> str:
        return f"turn_{uuid.uuid4().hex}"

    def phrase(self) -> str:
        return f"phrase_{uuid.uuid4().hex}"

class CounterIdSource:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._session = 0
        self._item = 0
        self._response = 0
        self._event = 0
        self._turn = 0
        self._phrase = 0

    def session(self) -> str:
        with self._lock:
            n = self._session
            self._session += 1
        return f"sess_{n:024d}"

    def item(self) -> str:
        with self._lock:
            n = self._item
            self._item += 1
        return f"item_{n:024d}"

    def response(self) -> str:
        with self._lock:
            n = self._response
            self._response += 1
        return f"resp_{n:024d}"

    def event(self) -> str:
        with self._lock:
            n = self._event
            self._event += 1
        return f"evt_{n:024d}"

    def turn(self) -> str:
        with self._lock:
            n = self._turn
            self._turn += 1
        return f"turn_{n:024d}"

    def phrase(self) -> str:
        with self._lock:
            n = self._phrase
            self._phrase += 1
        return f"phrase_{n:024d}"

_DEFAULT: IdSource = RandomIdSource()

def default_source() -> IdSource:
    return _DEFAULT

def next_session_id() -> str:
    return _DEFAULT.session()

def next_item_id() -> str:
    return _DEFAULT.item()

def next_response_id() -> str:
    return _DEFAULT.response()

def next_event_id() -> str:
    return _DEFAULT.event()

def next_turn_id() -> str:
    return _DEFAULT.turn()

def next_phrase_id() -> str:
    return _DEFAULT.phrase()

File: rust/src/ids.rs
#![allow(dead_code)]

use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;

use uuid::Uuid;

use super::types::{EventId, ItemId, ResponseId, SessionId};

pub trait IdSource: Send + Sync {
    fn session(&self) -> SessionId;
    fn item(&self) -> ItemId;
    fn response(&self) -> ResponseId;
    fn event(&self) -> EventId;
}

pub struct RandomIdSource;

impl IdSource for RandomIdSource {
    fn session(&self) -> SessionId {
        SessionId::new(format!("sess_{}", Uuid::new_v4().simple()))
    }
    fn item(&self) -> ItemId {
        ItemId::new(format!("item_{}", Uuid::new_v4().simple()))
    }
    fn response(&self) -> ResponseId {
        ResponseId::new(format!("resp_{}", Uuid::new_v4().simple()))
    }
    fn event(&self) -> EventId {
        EventId::new(format!("evt_{}", Uuid::new_v4().simple()))
    }
}

#[derive(Default)]
pub struct CounterIdSource {
    session: AtomicU64,
    item: AtomicU64,
    response: AtomicU64,
    event: AtomicU64,
}

impl CounterIdSource {
    pub fn new() -> Self {
        Self::default()
    }
}

impl IdSource for CounterIdSource {
    fn session(&self) -> SessionId {
        SessionId::new(format!(
            "sess_{:024}",
            self.session.fetch_add(1, Ordering::Relaxed)
        ))
    }
    fn item(&self) -> ItemId {
        ItemId::new(format!(
            "item_{:024}",
            self.item.fetch_add(1, Ordering::Relaxed)
        ))
    }
    fn response(&self) -> ResponseId {
        ResponseId::new(format!(
            "resp_{:024}",
            self.response.fetch_add(1, Ordering::Relaxed)
        ))
    }
    fn event(&self) -> EventId {
        EventId::new(format!(
            "evt_{:024}",
            self.event.fetch_add(1, Ordering::Relaxed)
        ))
    }
}

pub fn default_source() -> Arc<dyn IdSource> {
    Arc::new(RandomIdSource)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn random_ids_are_unique_and_prefixed() {
        let src = RandomIdSource;
        let a = src.item();
        let b = src.item();
        assert!(a.as_str().starts_with("item_"));
        assert!(b.as_str().starts_with("item_"));
        assert_ne!(a.as_str(), b.as_str());
        assert!(src.session().as_str().starts_with("sess_"));
        assert!(src.response().as_str().starts_with("resp_"));
        assert!(src.event().as_str().starts_with("evt_"));
    }

    #[test]
    fn counter_ids_are_deterministic() {
        let src = CounterIdSource::new();
        assert_eq!(src.item().as_str(), format!("item_{:024}", 0));
        assert_eq!(src.item().as_str(), format!("item_{:024}", 1));
        assert_eq!(src.response().as_str(), format!("resp_{:024}", 0));
        assert_eq!(src.session().as_str(), format!("sess_{:024}", 0));
        assert_eq!(src.event().as_str(), format!("evt_{:024}", 0));
    }

    #[test]
    fn counters_are_per_kind_independent() {
        let src = CounterIdSource::new();
        for _ in 0..3 {
            src.item();
        }
        assert_eq!(src.response().as_str(), format!("resp_{:024}", 0));
    }
}
